File: RobWorkSim/example/scripts/java/Simulation.java
import dk.robwork.*;
import java.util.Date;

public class Simulation {
	static RobWorkStudioPtr rwstudio;
	
	private static class Callback implements ThreadSimulatorStepCallbackHandler {
		long time;
		
		public Callback() {
			time = 0;
		}
		
		@Override
		public void callback(ThreadSimulator simulator, dk.robwork.State state) {
			Date now = new Date();
			if ((now.getTime()-time) > 1000) {
				time = now.getTime();
				rwstudio.setState(state);
				System.out.print("Time: ");
				System.out.println(simulator.getTime());
			}
		}
	}
	
	public static void main(String[] args) {
		// Use this for debugging:
		// System.out.println(System.getProperty("java.library.path"));
		
		// Load
		LoaderRW.load();
		LoaderRWS.load();
		LoaderRWSim.load();
		
		// Try to open RobWorkStudio and load the workcell
		rwstudio = rws.getRobWorkStudioInstance();
		DynamicWorkCellPtr dwc = DynamicWorkCellLoader.load("/home/thomas/Eclipse/LearnBiP/marvin/MarvinCalibratedScene.dwc.xml");
		WorkCellPtr wc = dwc.getWorkcell();
		dk.robwork.State state = wc.getDefaultState();
		rwstudio.postWorkCell(wc);
		
		// Create Simulator
		ContactDetectorPtr cd = new ContactDetectorPtr(new ContactDetector(wc));
		cd.setDefaultStrategies();
		PhysicsEnginePtr simulator = new PhysicsEnginePtr(new ODESimulator(dwc, cd));
		simulator.initPhysics(state);
		DynamicSimulatorPtr dsim = new DynamicSimulatorPtr(new DynamicSimulator(dwc, simulator));
		ThreadSimulatorPtr tsim = new ThreadSimulatorPtr(new ThreadSimulator(dsim, state));
		tsim.setTimeStep(0.001);
		tsim.setRealTimeScale(1);
		
		// Set target configuration for robot
		PDControllerPtr controller = dwc.findPDController("UR1Controller");
		Q qTarget = new Q(6,-1.666,-2.332,-0.413,-1.283,0.793,-0.219);
		controller.setTargetPos(qTarget);
		
		// Create callback
		ThreadSimulatorStepCallbackHandler cb = new Callback();
		ThreadSimulatorStepCallbackEnv fct = new ThreadSimulatorStepCallbackEnv(cb);
		tsim.setStepCallBack(fct);
		
		// Start simulation
		tsim.start();

		try {
			while (true) {
				Thread.sleep(1000);
				if (tsim.isInError())
					System.out.println("Simulation is in error!");
				if (!tsim.isRunning()) {
					System.out.println("Simulator stopped running!");
				}
			}
		} catch (InterruptedException e) {
			rwstudio.postExit();
		}
	}
}

File: RobWorkSim/example/scripts/matlab/simulation.m
%% Initialize
clear;
clc;

% Please see the following page for a description of how MatLab should be launched:
% http://www.robwork.dk/apidoc/nightly/rw/page_rw_scriptinterface.html

% Setup the paths (from the environment variables)
RW_LIBS=getenv('RW_LIBS');
RWSim_LIBS=getenv('RWSim_LIBS');
RWStudio_LIBS=getenv('RWStudio_LIBS');

% Import the java API
import dk.robwork.*;

% Load the native libraries
LoaderRW.load(RW_LIBS)
LoaderRWSim.load(RWSim_LIBS)
LoaderRWS.load(RWStudio_LIBS)

% Try to open RobWorkStudio and load the workcell
rwstudio = rws.getRobWorkStudioInstance();
dwc = DynamicWorkCellLoader.load('DynamicWorkCell.dwc.xml');
wc = dwc.getWorkcell();
state = wc.getDefaultState();
rwstudio.postWorkCell(wc);

% Create simulator
cd = ContactDetectorPtr(ContactDetector(wc));
cd.setDefaultStrategies();
simulator = PhysicsEnginePtr(ODESimulator(dwc, cd));
simulator.initPhysics(state);
dsim = DynamicSimulatorPtr(DynamicSimulator(dwc, simulator));
tsim = ThreadSimulatorPtr(ThreadSimulator(dsim, state));
tsim.setTimeStep(0.001);
tsim.setRealTimeScale(1);

% Set target configuration for robot
controller = dwc.findPDController('UR1Controller');
qTarget = Q(6,-1.666,-2.332,-0.413,-1.283,0.793,-0.219);
controller.setTargetPos(qTarget);

%% Create callback and run Simulation
dispatcher = ThreadSimulatorStepEventDispatcher();
setappdata(dispatcher,'UserData',[0]); % set step counter
% Set the callback (rwstudio pointer is passed as additional argument)
set(dispatcher,'StepEventCallback',{@StepCallBack,rwstudio});
% Manually invoke callback (just a test)
dispatcher.callback(tsim.dereference(),state);

% Add to simulator and start
fct = ThreadSimulatorStepCallbackEnv(dispatcher);
tsim.setStepCallBack(fct);
tsim.start();

%% Exit RobWorkStudio
if (tsim.isRunning())
    tsim.stop();
end
rwstudio.postExit();

